### PR TITLE
Integrate reservation system and normalization to decoding

### PR DESCRIPTION
**Pull request by rsuderman**
Normalization is significantly faster on device, so support whether the logits were returned in a normalized fashion when doing beam search. This allows existing support for previous decoding strategies.

Also add support for decoder reservation system to better support batched decoding.

### CODE REVIEW

**Review by stbaione — APPROVED** [review]
LGTM, the `config.py` file really cleans things up

It looks like the changes to `Batcher` are slowing down, so I can take a look at #1195 when I have `sampling` in a good spot
